src/api/uploads/MultiputUpload.js

```diff
@@ -0,0 +1,358 @@
+    digestReadahead: 5, // How many parts past those currently uploading to precompute digest for
+    initialRetryDelayMs: 5000, // Base for exponential backoff on retries
+    maxRetryDelayMs: 60000, // Upper bound for time between retries
+    parallelism: 5, // Maximum number of parts to upload at a time
+    requestTimeout: 120000, // Idle timeout on part upload, overall request timeout on other requests
+    // eslint-disable-next-line max-len
+    retries: 5 // How many times to retry requests such as upload part or commit. Note that total number of attempts will be retries + 1 in worst case where all attempts fail.
+};
+
+class MultiputUpload extends Base {
+    clientId: ?string;
+    commitRetryCount: number;
+    config: Object;
+    createSessionNumRetriesPerformed: number;
+    destinationFile: ?string;
+    destinationFolder: ?string;
+    file: File;
+    fileSha1: ?string;
+    firstUnuploadedPartIndex: number;
+    initialFileLastModified: ?string;
+    initialFileSize: number;
+    onCompleted: ?Function;
+    onFailure: ?Function;
+    onProgress: ?Function;
+    options: Object;
+    partSize: number;
+    parts: Array<MultiputPart>;
+    partsDigestComputing: number;
+    partsDigestReady: number;
+    partsNotStarted: number;
+    partsUploaded: number;
+    partsUploading: number;
+    sessionEndpoints: Object;
+    sessionId: ?string;
+    totalUploadedBytes: number;
+    worker: Worker;
+
+    /**
+     * [constructor]
+     * 
+     * @param {object} options
+     * @param {File} file
+     * @param {string} createSessionUrl
+     * @param {string} [destinationFolder] - Untyped folder id (e.g. no "d_" prefix)
+     * @param {string} [destinationFile] - Untyped file id (e.g. no "f_" prefix)
+     * @param {MultiputConfig} [config]
+     */
+    constructor(
+        options: Object,
+        file: File,
+        createSessionUrl: string,
+        destinationFolder?: ?string,
+        destinationFile?: ?string,
+        config?: MultiputConfig
+    ) {
+        super(options);
+
+        if (destinationFolder !== null && destinationFile !== null) {
+            throw new Error('Both destinationFolder and destinationFile set');
+        }
+        this.file = file;
+
+        // These values are used as part of our (best effort) attempt to abort uploads if we detect
+        // a file change during the upload.
+        this.initialFileSize = this.file.size;
+        this.initialFileLastModified = getFileLastModifiedAsISONoMSIfPossible(this.file);
+
+        this.sessionEndpoints = {
+            createSession: createSessionUrl,
+            uploadPart: null,
+            listParts: null,
+            commit: null,
+            abort: null,
+            logEvent: null
+        };
+        // Explicitly cast these two to string because sometimes they get passed as int, and
+        // sending folder_id as a number instead of a string in the JSON body of createSession
+        // results in an API error.
```

Comment:
Looks like this comment may be out of date?

Reply:
ah removed